compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6256,17 +6256,7 @@ private void checkStringTemplateExprs(List<? extends BLangExpression> exprs, boo
                 continue;
             }
 
-            if (type.tag >= TypeTags.JSON) {
-                if (allowXml) {
-                    if (type.tag != TypeTags.XML) {
-                        dlog.error(expr.pos, DiagnosticErrorCode.INCOMPATIBLE_TYPES,
-                                BUnionType.create(null, symTable.intType, symTable.floatType,
-                                        symTable.decimalType, symTable.stringType,
-                                        symTable.booleanType, symTable.xmlType), type);
-                    }
-                    continue;
-                }
-
+            if (!types.isUnionOfSimpleBasicTypesExceptNilType(type)) {
                 dlog.error(expr.pos, DiagnosticErrorCode.INCOMPATIBLE_TYPES,
```

Comment:
Error message says `expected '(int|float|decimal|string|boolean)'` but actually we expect a subtype. Propose to have a new message.

Reply:
On second thoughts, this is like any other mismatched type error we have, right? WDYT?

```ballerina
function func(int|float|decimal|boolean|string x) {
}

public function main() {
    int[] x = [];
    func(x); // [ERROR] incompatible types: expected '(int|float|decimal|boolean|string)', found 'int[]'
}
```